streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java

```diff
@@ -334,10 +335,33 @@ public void close(final Map<TopicPartition, Long> offsets) throws IOException {
 
     @Override
     public void checkpoint(final Map<TopicPartition, Long> offsets) {
+
+        // Find non persistent store's topics
+        Map<String, String> storeToChangelogTopic = topology.storeToChangelogTopic();
+        Set<String> globalNonPersistentStoresTopics = new HashSet<>();
+        for (StateStore store : topology.globalStateStores()) {
+            if (!store.persistent() && storeToChangelogTopic.containsKey(store.name())) {
+                globalNonPersistentStoresTopics.add(storeToChangelogTopic.get(store.name()));
+            }
+        }
+
         checkpointableOffsets.putAll(offsets);
```

Comment:
Should we not filter first before we put the offset into `checkpointableOffsets` ?

Reply:
Actually the returned map from `GlobalStateUpdateTask#initialize` will be used for the global thread to assign partitions and seek positions to the consumer, so for non-persistent stores we should set the entry as `topic-partition: NO_CHECKPOINT` where `NO_CHECKPOINT` is defined in `StateRestorer` as -1. Then in the thread it should handle `NO_CHECKPOINT` by seeking to beginning.

You can take a look at this part of the code for non-global restoration logic as a reference: https://github.com/apache/kafka/blob/27b56b1458d0af0a397e9f47ecd008c10c088c6d/streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java#L167-L184

Reply:
Would you please check my latest commit @guozhangwang. I hope that I got the idea.

Reply:
By the way I noticed that checkpoint file is deleted under trunk (1.2.x) when shut down occurs. But not under 1.0.x. This might be another bug? 

I have also added another PR at [here] targetting 1.0.x . This is actually bug which affected us. 1.2.x is deleting file and we are not affected somehow.  

[here]: https://github.com/apache/kafka/pull/4789

Reply:
In 1.2.0 we add an optimization to avoid writing the checkpoint file if there is nothing to write (i.e. the available offset map is empty): this is not a bug fix but just some optimization. If you have other persistent stores in your topology the checkpoint file will still be written. Here is the JIRA ticket: https://issues.apache.org/jira/browse/KAFKA-6499